csharp/src/Microsoft.ML.OnnxRuntime/NativeMethods.shared.cs

```diff
@@ -890,8 +892,8 @@ internal class NativeLib
         /// <param name="configValue">Config value</param>
         [UnmanagedFunctionPointer(CallingConvention.Winapi)]
         public delegate IntPtr /*(OrtStatus*)*/ DOrtAddSessionConfigEntry(IntPtr /* OrtSessionOptions* */ options,
-                                                                          IntPtr /* const char* */configKey,
-                                                                          IntPtr /* const char* */ configValue);
+                                                                          byte[] /* const char* */configKey,
+                                                                          byte[] /* const char* */ configValue);
```

Comment:
The managed memory must be pinned before being passed to the native code so it is not GCed or moved. Before this PR we relied on GCHandle like pinning (same as MemoryHandle) which is very expensive as it turns out. The passed IntPtr did not require marshalling.

I learned that when one passes blittable array of blittable types it is pinned automatically using 'fixed()' style much more efficient pinning. And we do not have to do try/finally for that.  We sort of used it in other places if one looks carefully (IntPtr[] is also considered blittable).

In fact, this was one of the most expensive parts.

Reply:
Is there a guideline that could be documented so that going forward the best thing to do can be captured in that instead of trying to infer it from existing implementation details?